### PR TITLE
Convert babelfishpg_tsql to not use direct links to babelfishpg_common

### PR DESCRIPTION
**Pull request by rishabhtanwar29**
### Description
Historically, we have referenced functions from `babelfishpg_common` directly
into `babelfishpg_tsql` extension using `extern`keyword. Due to this `babelfishpg_tsql`
library couldn't be loaded independently (without explicitly loading `babelfishpg_common`
library first) due to unresolved/undefined symbols. To overcome this commit 06fdc75bfbb9ef8be16041ea052526ff761bea87
linked `babelfishpg_common` to `babelfishpg_tsql` with a direct dependency.

The direct linking fix mentioned above has it's own drawbacks and causes MVU failures
due to incorrect library linking between source and target data directories during MVU.
Instead, let's remove these `extern` references (unresolved symbols) and use function
pointers that are manually filled in by the module's _PG_init function and all are clubbed
together into one `rendezvous` variable called `common_utility_plugin`.
Additionally, change all occurrences of PG directory to `psql` so that directory names
do not contain `postgres` or `pgsql` keywords.

**Note**: I have temporarily disabled all upgrade workflows (except latest->latest) since
MVU to PG 15 is blocked on BABEL-3801 (which will be fixed soon). These workflows
will be enabled back with the fix for BABEL-3801 but I have manually tested these changes
and verified that MVU works fine.

Task: BABEL-3758
Signed-off-by: Rishabh Tanwar <ritanwar@amazon.com>

### Test Scenarios Covered ###
* **Use case based -**

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: .github/composite-actions/build-extensions/action.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@ inputs:
   install_dir:
     description: 'Engine install directory'
     required: no
-    default: postgres
+    default: psql
```
Comment (surendravishnoi) [inline .github/composite-actions/build-extensions/action.yml:6]: why is this change required? if keywords are issue, it would be better to use something other than psql

File: .github/workflows/major-version-upgrade.yml [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
 jobs:
   run-babelfish-mvu-tests:
     env:
-      OLD_INSTALL_DIR: postgres13
-      NEW_INSTALL_DIR: postgres14
-      ENGINE_BRANCH_FROM: BABEL_1_X_DEV__PG_13_X
-      EXTENSION_BRANCH_FROM: BABEL_1_X_DEV
+      OLD_INSTALL_DIR: psql_source
+      NEW_INSTALL_DIR: psql_target
+      ENGINE_BRANCH_FROM: BABEL_2_3_STABLE__PG_14_6
```
Comment (surendravishnoi) [inline .github/workflows/major-version-upgrade.yml:9]: why is this 14_6? is it from 14_6 to 15 latest for empty database?

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 3]
```diff
@@ -155,7 +154,7 @@ version(PG_FUNCTION_ARGS)
 	 * TODO: Return Build number with version string as well.
 	 */
 
-	info = tsql_varchar_input(temp.data, temp.len, -1);
+	info = (*common_utility_plugin_ptr->tsql_varchar_input)(temp.data, temp.len, -1);
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/runtime/functions.c:157]: how are we ensuring that common_utility_plugin_ptr would always have initialised at this point?

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 4]
```diff
@@ -236,7 +235,7 @@ InstallExtendedHooks(void)
 	logicalrep_modify_slot_hook = logicalrep_modify_slot;
 
 	prev_is_tsql_rowversion_or_timestamp_datatype_hook = is_tsql_rowversion_or_timestamp_datatype_hook;
-	is_tsql_rowversion_or_timestamp_datatype_hook = is_tsql_rowversion_or_timestamp_datatype;
+	is_tsql_rowversion_or_timestamp_datatype_hook = common_utility_plugin_ptr->is_tsql_rowversion_or_timestamp_datatype;
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c:240]: we should have init'd this hook in common extension side.

File: contrib/babelfishpg_common/src/datetime.h [hunk 5]
```diff
@@ -8,6 +8,10 @@
 #ifndef PLTSQL_DATETIME_H
 #define PLTSQL_DATETIME_H
 
+#include "postgres.h"
+
+#include "datatype/timestamp.h"
```
Comment (surendravishnoi) [inline contrib/babelfishpg_common/src/datetime.h:11]: avoid including headers in headers, rather include in C files wherever needed (OK for cases where header file itself needs the def)

File: contrib/babelfishpg_common/src/sqlvariant.h [hunk 6]
```diff
@@ -1,3 +1,10 @@
+#ifndef SQLVARIANT_H
+#define SQLVARIANT_H
+
+#include "postgres.h"
```
Comment (surendravishnoi) [inline contrib/babelfishpg_common/src/sqlvariant.h]: avoid headers inclusion in header, include in C files wherever needed

File: contrib/babelfishpg_tsql/runtime/functions.c (hunk 3 above)
Comment (rishabhtanwar29) [reply contrib/babelfishpg_tsql/runtime/functions.c:157]: It is getting inited in `_pg_init` so it will always be initialized as we have an explicit load_library('babelfishpg_common') in `_pg_init`. For reference see [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/910/files#diff-ce1101bab57d0e7f8b36f0df103ed3ae49f4b488e473ac87bf2320c32f2697e1R3348).

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 7]
```diff
@@ -889,3 +890,19 @@ get_pltsql_function_signature(PG_FUNCTION_ARGS)
...
+	{
+		common_utility_plugin **utility_plugin;
+		utility_plugin = (common_utility_plugin **) find_rendezvous_variable("common_utility_plugin"); 
+		common_utility_plugin_ptr = *utility_plugin;
+
+		/* common_utility_plugin_ptr is still not initialised */
+		if (!common_utility_plugin_ptr)
+			ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("common utility plugin pointer is not initialised properly.")));
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: Give a better error like "Failed to find common utility plugin"

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 4 above)
Comment (rishabhtanwar29) [reply contrib/babelfishpg_tsql/src/hooks.c:240]: We will not get anything extra even if we do that since the specific plugin (`is_tsql_rowversion_or_timestamp_datatype`) is anyway needed in tsql extensions in other places as well so it's not like we are reducing one function from the structure. But I will address this separately.

File: contrib/babelfishpg_tsql/src/pltsql.h [hunk 8]
```diff
@@ -34,6 +34,7 @@
 
 #include "dynavec.h"
 #include "dynastack.h"
+#include "../../contrib/babelfishpg_common/src/babelfishpg_common.h"
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql.h:37]: why is this required? We already have callbacks from TSQL to common (for collation etc.), how do they work?

File: .github/composite-actions/build-extensions/action.yml (hunk 1 above)
Comment (rishabhtanwar29) [reply .github/composite-actions/build-extensions/action.yml:6]: Problematic keywords are `postgres` and `pgsql` only so `psql` will work fine but let me know if you have any preference.

File: .github/workflows/major-version-upgrade.yml (hunk 2 above)
Comment (rishabhtanwar29) [reply .github/workflows/major-version-upgrade.yml:9]: As we are not supporting MMVU so I am changing it as 14_6 to 15 latest MVU and yes it is for empty databases.

File: contrib/babelfishpg_common/src/datetime.h (hunk 5 above)
Comment (rishabhtanwar29) [reply contrib/babelfishpg_common/src/datetime.h:11]: It is needed in the header file itself (there is a reference for `Timestamp` in this file).

**Review by Deepesh125 — DISMISSED** [review]

File: contrib/babelfishpg_tsql/src/pltsql.h (hunk 8 above)
Comment (rishabhtanwar29) [reply contrib/babelfishpg_tsql/src/pltsql.h:37]: This is needed so that we don't have to declare the rendezvous variable structure at two places. Other examples [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/BABEL_3_X_DEV/contrib/babelfishpg_tds/src/include/tds_int.h#L37). For collation callbacks, we have declared the structure both in tsql and common extensions which is just duplicate code so I think it is better to keep the structure at one place and use this header include method to reduce duplication of the code.

File: contrib/babelfishpg_tsql/src/pltsql_utils.c (hunk 7 above)
Comment (rishabhtanwar29) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c]: Updated!

File: contrib/babelfishpg_tsql/src/pltsql.h (hunk 8 above)
Comment (surendravishnoi) [reply contrib/babelfishpg_tsql/src/pltsql.h:37]: fine for now but we need to think more. We have taken care of link time dependency but added a compile time dependency from TSQL extension to common extension. I guess such exchange should pass thorough engine as that is common to all extensions.

**Review by surendravishnoi — DISMISSED** [review]

**Review by Deepesh125 — DISMISSED** [review]

**Review by Deepesh125 — DISMISSED** [review]

**Review by Deepesh125 — APPROVED** [review]
